### ISSUE TITLE
GPTNeoX generate won't work

### ISSUE THREAD

**Issue by shivance** [keras-team/keras-nlp#1183]
Following up the discussion from #1110 
`generate()` with GPTNeoXCausalLM with pythia-70m checkpoints is not working

**Comment by shivance** [comment]
@mattdangerw update on this front..
Here is a [colab](https://colab.research.google.com/drive/1OatD7MaCufZR2je8O1UOWEyR8YPoe_rT?usp=sharing) that compares the output of our new multi backend GPTNeoXBackbone with that of HF. 

They match till third decimal point.

**Comment by abheesht17** [comment]
@shivance - can you please check whether `"gelu_new"` and `"gelu_fast"` are numerically equivalent?

GPT-2 uses `"gelu_new"` (I think this is same as `keras.activations.gelu(..., approximate=True`), GPT-NeoX uses `"gelu_fast"`.

https://huggingface.co/gpt2/blob/main/config.json#L2
https://huggingface.co/EleutherAI/gpt-neox-20b/blob/main/config.json#L8

**Comment by shivance** [comment]
https://github.com/huggingface/transformers/blob/05cda5df3405e6a2ee4ecf8f7e1b2300ebda472e/src/transformers/activations.py#L49

https://github.com/huggingface/transformers/blob/05cda5df3405e6a2ee4ecf8f7e1b2300ebda472e/src/transformers/activations.py#L81

I think it's game of approximation

**Comment by abheesht17** [comment]
A couple of other things you can try...

- Check whether your preprocessed input has the end token or not. If it does, then further generation will not take place. This might be why your input = output.
- You can try generating w/o caching first.

**Comment by shivance** [comment]
> @shivance - can you please check whether `"gelu_new"` and `"gelu_fast"` are numerically equivalent?
> 
> GPT-2 uses `"gelu_new"` (I think this is same as `keras.activations.gelu(..., approximate=True`), GPT-NeoX uses `"gelu_fast"`. There might be numerical differences here.
> 
> https://huggingface.co/gpt2/blob/main/config.json#L2 https://huggingface.co/EleutherAI/gpt-neox-20b/blob/main/config.json#L8

Great find ! I think the cause of precision mismatch is the activation function itself.

**Comment by shivance** [comment]
@mattdangerw something is terribly wrong with generate pipeline
1. As said by @abheesht17 I passed preprocessed tokens to `generate()` and it gives output.
2. The tokens in output seem to have been randomly picked !
3. detokenization of generated tokens is not being performed correctly -> generate returns input string while manually detokenizing the output gives something else which looks like this
4. Last but not least, for `pythia-70m-deduped` checkpoint, in that random output only two words repeat, atleast it's more random with `pythia-140m-deduped`

<img width="1092" alt="Screenshot 2023-07-30 at 4 58 29 PM" src="https://github.com/keras-team/keras-nlp/assets/51750587/0cd4798e-5235-4968-a918-e2a1dd49c479">

**Comment by mattdangerw** [comment]
@shivance yeah, the generation graph is quite a step up in complexity from just aligning the forward pass, so makes sense we might have some bugs here.

- We might still have some issue with `generate_step` in the GptNeoXCausalLM.
- Another thing I suspect we might be missing is the equivalent to [the start index](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/modeling/position_embedding.py#L99) in `PositionEmbedding`. Basically to get generation working correctly for a fixed positional embedding, we need to use the current index of our generation loop. Do we need to do something similar for rotary position embeddings?

I would suggest trying to triangulate this in eager mode (either tf or torch), and work against a known reference implementation. This might be tricky to set up, but will save time overall!

**Comment by shivance** [comment]
> * Another thing I suspect we might be missing is the equivalent to [the start index](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/modeling/position_embedding.py#L99) in `PositionEmbedding`. Basically to get generation working correctly for a fixed positional embedding, we need to use the current index of our generation loop. Do we need to do something similar for rotary position embeddings?

The inability to generate is unrelated to start_index, I forgot to update here, but last week I created a branch on my fork to add start_index and corresponding code snippet for rotary embedding (same as #1209), it doesn't fix it.

**Comment by maitry63** [comment]
Hi @kanpuriyanawab, based on the [Official KerasHub GPT-NeoX API Documentation](https://keras.io/keras_hub/api/models/gpt_neo_x/gpt_neo_x_backbone/), the small difference in outputs between `KerasHub` and `Hugging Face` is completely normal. 

It is caused by how the two frameworks are designed, so it is not a bug.

The main reasons for this difference are:

- `RoPE Slicing Dimensions`: When rotating a part of the hidden states (rotary_percentage=0.25), KerasHub pairs and mixes adjacent odd and even numbers together. 
On the other hand, reference libraries cut the numbers into two straight halves before rotating them.

- `Parallel Residual`: KerasHub runs the Attention and MLP parts side-by-side using an automatic graph setup, instead of adding them together step-by-step.

I tried to unpack Hugging Face's multi-head QKV weights and manually rearrange them to fit KerasHub's layout layer by layer. 

Even though the model weights now line up perfectly, the hardcoded math inside each framework explains why the final outputs are slightly different.

You can check this [gist](https://colab.sandbox.google.com/gist/maitry63/2f716262339f9e1d1889bebb7853e655/output_matching_merged_gpt.ipynb#scrollTo=x1s67mkR0MGz) to see the exact code and results.
